# 개요

- 실전 데이터를 이용하여 데이터 준비 과정 체크

# 연구 목표 : 요구사항 분석

- 주제
  - 고객, 상품, 주문정보에 대한 분석
  - 이를 통한 정량적 분석
- 목표
  - pandas 연습

# 데이터 수집(제공)

- level 1 : 사내 데이터 제공
- 내역 (csv)
  - 고객 마스터 데이터 1개
  - 아이템(상품) 마스터 데이터 1개
  - 주문 정보 2개
  - 주문 상세 정보 2개
  - db 구조에 맞춰 데이터를 추출하여 제공한것(가정)

# 데이터 준비

### step1 데이터(csv)를 df로 로드한다

In [114]:
import numpy as np
import pandas as pd

# 아래처럼 관리 혹은 os 모듈(경로를 운영체계에 맞게 동적 구성) 활용
# 단, 코랩은 무조건 리눅스
csv_dir = '/content/drive/MyDrive/2. 데이터분석/data/'
df1 = pd.read_csv(csv_dir+'customer_master.csv')
df2 = pd.read_csv(csv_dir+'transaction_1.csv')
df3 = pd.read_csv(csv_dir+'transaction_2.csv')
df4 = pd.read_csv(csv_dir+'transaction_detail_1.csv')
df5 = pd.read_csv(csv_dir+'transaction_detail_2.csv')
df6 = pd.read_csv(csv_dir+'item_master.csv')

In [115]:
# temp.head(1)
# 결측이 많다 => 컬럼이 다른 형태가 많다
# 같은 컬럼을 가진 데이터간에는 단순 합치기
# 다른 컬럼을 가진 데이터간에는 머지(조인) 진행

In [116]:
# temp = pd.concat([df1, df2, df3, df4, df5, df6])
# temp.shape

In [117]:
# 실습
# customer_master.csv -> df, item_master.csv->df,...
# dfs라는 리스트에 [df, df, df, df, df, df]로 로드 시키시오

# 해당 파일 목록 획득(특정 폴더 밑에 필요한 데이터 배치했다는 가정하에 진행)
import glob
# 일일이 작성하기 귀찮을 수도 (파일이 엄청 많다면) 진행
# 단, 반드시 생성된 df의 순서를 알고 있어야 한다
files = glob.glob('/content/drive/MyDrive/2. 데이터분석/data/*.csv')
files.sort() # 순서를 일정하게 고정
files

['/content/drive/MyDrive/2. 데이터분석/data/customer_master.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/item_master.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/tr.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/transaction_1.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/transaction_2.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/transaction_detail_1.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/transaction_detail_2.csv',
 '/content/drive/MyDrive/2. 데이터분석/data/중간결과물.csv']

In [118]:
# 리스트 컴프리핸션으로 구성 -> 덤으로 연습
# dfs 생성
# csv -> df
dfs = [pd.read_csv(file) for file in files]
dfs

[     customer_id customer_name registration_date                        email  \
 0       IK152942           김서준   2019-01-01 0:25  hirata_yuujirou@example.com   
 1       TS808488           김예준   2019-01-01 1:13    tamura_shiori@example.com   
 2       AS834628           김도윤   2019-01-01 2:00      hisano_yuki@example.com   
 3       AS345469           김시우   2019-01-01 4:48   tsuruoka_kaoru@example.com   
 4       GD892565           김주원   2019-01-01 4:54   oouchi_takashi@example.com   
 ...          ...           ...               ...                          ...   
 4995    AS677229           정우석  2019-07-31 16:52    hirayama_risa@example.com   
 4996    HD758694           정영훈  2019-07-31 19:09  nakahara_mahiru@example.com   
 4997    PL538517           정준기  2019-07-31 19:30      tabata_yuu1@example.com   
 4998    OA955088           정도형  2019-07-31 22:32  setouchi_hikaru@example.com   
 4999    HI349563           정지석  2019-07-31 22:49      horii_kanji@example.com   
 
      gender  

In [119]:
dfs = list()
# csv를 하나씩 뽑아서
for file in files:
  # csv -> df로 구성 -> 리스트에 담는다(요구사항)
  dfs.append(pd.read_csv(file))
dfs

[     customer_id customer_name registration_date                        email  \
 0       IK152942           김서준   2019-01-01 0:25  hirata_yuujirou@example.com   
 1       TS808488           김예준   2019-01-01 1:13    tamura_shiori@example.com   
 2       AS834628           김도윤   2019-01-01 2:00      hisano_yuki@example.com   
 3       AS345469           김시우   2019-01-01 4:48   tsuruoka_kaoru@example.com   
 4       GD892565           김주원   2019-01-01 4:54   oouchi_takashi@example.com   
 ...          ...           ...               ...                          ...   
 4995    AS677229           정우석  2019-07-31 16:52    hirayama_risa@example.com   
 4996    HD758694           정영훈  2019-07-31 19:09  nakahara_mahiru@example.com   
 4997    PL538517           정준기  2019-07-31 19:30      tabata_yuu1@example.com   
 4998    OA955088           정도형  2019-07-31 22:32  setouchi_hikaru@example.com   
 4999    HI349563           정지석  2019-07-31 22:49      horii_kanji@example.com   
 
      gender  

In [120]:
len(dfs)

8

#### 기초점검

In [121]:
# 고객 마스터 정보 -> 0번
dfs[0].shape, display(dfs[0].head(1))

customer_id customer_name registration_date                        email  \
0    IK152942           김서준   2019-01-01 0:25  hirata_yuujirou@example.com   

  gender  age       birth   pref  
0      M   29  1990-06-10  대전광역시

((5000, 8), None)

In [122]:
# 아이템 마스터 정보 -> 5번, (5,3), 상품 자체는 파악 불가능함, 상품으로 인지
dfs[1].shape, display(dfs[5].head(1))

detail_id transaction_id item_id  quantity
0          0    T0000000113    S005         1

((5, 3), None)

In [123]:
# 원래 5000 + 1786 => 5천개씩 끊어서 제공, 개별 데이터 별로 인덱스 값 0 출발
# payment_date 날짜가 연속적
# 트랜젝션 정보 -> 2 (5000,4)
dfs[2].shape, display(dfs[2].head(1))

Unnamed: 0  detail_id transaction_id item_id  quantity   price  \
0           0          0    T0000000113    S005         1  210000   

          payment_date customer_id item_name  item_price  
0  2019-02-01 01:36:57    PL563502      PC-E      210000

((7144, 10), None)

In [124]:
dfs[2].shape, display(dfs[2].head(1))

Unnamed: 0  detail_id transaction_id item_id  quantity   price  \
0           0          0    T0000000113    S005         1  210000   

          payment_date customer_id item_name  item_price  
0  2019-02-01 01:36:57    PL563502      PC-E      210000

((7144, 10), None)

In [125]:
dfs[2].tail(2)
# 트랜젝션 1 + 트랜젝션 2 => 시간의 순서대로 데이터가 배치된다 => 병합 (단순합치기, 데이터증가)

Unnamed: 0  detail_id transaction_id item_id  quantity  price  \
7142        7142       7142    T0000006897    S002         1  85000   
7143        7143       7143    T0000006898    S002         1  85000   

             payment_date customer_id item_name  item_price  
7142  2019-07-31 23:39:35    TS624738      PC-B       85000  
7143  2019-07-31 23:41:38    AS834214      PC-B       85000

In [126]:
# 트랜젝션 정보 -> 3 (1786,4)
# 가격, 결제시간, 고객정보
# 트랜젝션, 고객 마스터 => merge 작업 필요 => 1개로 병합
dfs[3].shape, display(dfs[3].head(1))

transaction_id   price         payment_date customer_id
0    T0000000113  210000  2019-02-01 01:36:57    PL563502

((5000, 4), None)

- 트랜젝션 정보
  - 총 데이터는 5000 + 1786
  - 인덱스 0 ~ 4999, 0 ~ 1785
  - 구조가 동일하므로 -> concat() 수행
  - 합치는 순서
    - 파일명이 1, 2이므로 순서
    - payment_date 날짜 확인 시간의 순으로 합치기 -> 1, 2 순임
  - 컬럼
    - transaction_id : 고유한 거래 1건에 대한 구분값
    - price : 매출액
    - payment_date : 결재시간
    - customer_id : 고객번호

In [127]:
# 트랜젝션 디테일 정보 -> 4, 5
# 상세정보 : 제품번호, 수량
# 트랜젝션, 트랜젝션 디테일 => merge 작업 필요 => 1개로 병합
# 트랜젝션 디테일, 아이템 마스터 => merge 작업 필요 => 1개로 병합
dfs[4].shape, display(dfs[4].head(1))

transaction_id   price         payment_date customer_id
0    T0000005113  295000  2019-06-15 07:20:27    TS169261

((1786, 4), None)

In [128]:
dfs[4].tail(1)

transaction_id  price         payment_date customer_id
1785    T0000006898  85000  2019-07-31 23:41:38    AS834214

In [129]:
# 트랜젝션 디테일 1 + 트랜젝션 디테일 2 => concat 병합, 데이터가 늘어나는 방향
dfs[5].head(3)

detail_id transaction_id item_id  quantity
0          0    T0000000113    S005         1
1          1    T0000000114    S001         1
2          2    T0000000115    S003         1

In [130]:
dfs[5].shape

(5000, 4)

- 트랜젝션 디테일
  - 총 데이터는 5000+2144
    - 트랜젝션과 크기가 다름
    - 트랜젝션 디테일이 더 많다
      - 트랜젝션 데이터가 누락? x
      - 트랜젝션 데이터 1개의 디테일이 2개이상 적용(연결) 될 수 있다 o
        - transaction_id가 중복된것이 존재한다면?

In [131]:
dfs[5].transaction_id.value_counts()
# 이 결과로 트랜젝션 디테일이 n개가 1개의 트랜젝션과 연관된것이 존재
# 나중에 병합할때 거래의 총 개수는 트랜젝션을 기준(5000+1786)으로 진행
# 트랜젝션 유지, 트랜젝션 디테일을 연결 => left join(한가지 생각)
# 반대 관점 디테일 중심 트랜젝션 연결(디테일은 LEFT JOIN, 5000+2144) => 모든 디테일 정보 노출(건건이)
# 단, 트랜젝션은 중복적으로 등장 가능함

transaction_id
T0000000790    4
T0000002924    3
T0000000397    3
T0000004784    3
T0000000683    3
              ..
T0000001734    1
T0000001733    1
T0000001731    1
T0000001730    1
T0000001740    1
Name: count, Length: 4757, dtype: int64

In [132]:
# transaction_id T0000005335 번에 대해 트랜잭션, 트랜잭션 디테일 데이터를
# 각각 구하시오, 실습

# 트랜잭션 데이터 2, 3
dfs[3][ dfs[3].transaction_id == 'T0000005335' ]

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id]
Index: []

In [133]:
# 트랜젝션 디테일 데이터 4, 5
dfs[5][ dfs[5].transaction_id == 'T0000005335' ]
# 디테일: 제품 1건에 대해서 개별 데이터(로그) 이해

Empty DataFrame
Columns: [detail_id, transaction_id, item_id, quantity]
Index: []

In [134]:
# 트랜젝션 1개에 디테일 3개가 머지 되면? ->어떻게 노출되는지 확인 필요!!
# 좀더 진행후 확인

- 결론
  - 1개의 df로 병합 가능
  - 트랜젝션 1 + 트랜젝션 2 => concat, axis = 0 (세로방향) => 데이터가 증가
  - 트랜젝션-디 1 + 트랜젝션-디 2 => concat, axis = 0 (세로방향) => 데이터가 증가
  - 트렌젝션 + 트렌젝션-디 => merge, inner join => 트랜젹션 번호로 합치기
    - 데이터의 수 = 5000 + 2144 = 트랜젝션의 총수(7144)
  - 트렌젝션 + 고객 마스터 => merge, left join => 고객번호로 합치기
  - 트렌젝션-디 + 아이팀 마스터 => merge, left join => 아이템 번호로 합치기


  - 예상되는 데이터 shape
    - (5000 + 1786, ?)
    - 왼쪽에 위치된 데이터에 따라 달라질수 있다

### step2 병합 -> 1개의 df로 구성

#### 트랜젝션 1 + 트랜젝션 2 => concat

In [135]:
# 실습 2분
transaction_1 = dfs[2]
transaction_2 = dfs[3]

transaction = pd.concat([dfs[2], dfs[3]],axis=0)
transaction.shape

(12144, 10)

In [136]:
# 합쳐진 지점 확인 -> 인덱스 부분, 개별 데이터는 모두 인덱스가 0으로 출발했음
transaction [5000-2:5000+2] # 합쳐지는 지점 확인

# 인덱스 값이 로테이션 => 무시하고 새로 세팅하던지(reset), 초기화하던지(생성시 세팅)

Unnamed: 0  detail_id transaction_id item_id  quantity   price  \
4998      4998.0     4998.0    T0000004868    S005       1.0  210000   
4999      4999.0     4999.0    T0000004869    S003       1.0  120000   
5000      5000.0     5000.0    T0000004870    S002       3.0  255000   
5001      5001.0     5001.0    T0000004871    S003       1.0  120000   

             payment_date customer_id item_name  item_price  
4998  2019-06-08 17:17:20    PL613382      PC-E    210000.0  
4999  2019-06-08 17:18:18    HD502657      PC-C    120000.0  
5000  2019-06-08 17:57:05    IK514038      PC-B     85000.0  
5001  2019-06-08 19:53:49    GD574697      PC-C    120000.0

In [137]:
# ignore_index=True : 기존 원 데이터에 인덱스 무시
transaction = pd.concat([dfs[2], dfs[3]],axis=0, ignore_index=True)
transaction.shape

(12144, 10)

In [138]:
# 데이터가 합쳐지는 부분 확인
transaction[ 5000-2:5000+2 ]

Unnamed: 0  detail_id transaction_id item_id  quantity   price  \
4998      4998.0     4998.0    T0000004868    S005       1.0  210000   
4999      4999.0     4999.0    T0000004869    S003       1.0  120000   
5000      5000.0     5000.0    T0000004870    S002       3.0  255000   
5001      5001.0     5001.0    T0000004871    S003       1.0  120000   

             payment_date customer_id item_name  item_price  
4998  2019-06-08 17:17:20    PL613382      PC-E    210000.0  
4999  2019-06-08 17:18:18    HD502657      PC-C    120000.0  
5000  2019-06-08 17:57:05    IK514038      PC-B     85000.0  
5001  2019-06-08 19:53:49    GD574697      PC-C    120000.0

In [139]:
# 마지막 데이터 확인
transaction.tail(2)

Unnamed: 0  detail_id transaction_id item_id  quantity   price  \
12142         NaN        NaN    T0000005111     NaN       NaN  210000   
12143         NaN        NaN    T0000005112     NaN       NaN   50000   

              payment_date customer_id item_name  item_price  
12142  2019-06-15 04:14:06    IK074758       NaN         NaN  
12143  2019-06-15 04:42:38    HD444151       NaN         NaN

#### 트랜젝션-디 1 + 트랜젝션-디 2 => concat

In [140]:
# 실
transaction_detail = pd.concat([dfs[4], dfs[5]], axis=0, ignore_index=True)
transaction_detail.shape

# 트랜젝션 6786, 디테일 7144 => 병합후 확인 필요

(6786, 7)

In [141]:
transaction_detail.sample(5)

transaction_id     price         payment_date customer_id  detail_id  \
3313    T0000001563       NaN                  NaN         NaN     1527.0   
1795    T0000000122       NaN                  NaN         NaN        9.0   
2105    T0000000418       NaN                  NaN         NaN      319.0   
1588    T0000006701  100000.0  2019-07-27 00:20:05    OA923215        NaN   
2185    T0000000495       NaN                  NaN         NaN      399.0   

     item_id  quantity  
3313    S003       1.0  
1795    S002       1.0  
2105    S004       1.0  
1588     NaN       NaN  
2185    S001       1.0

#### 트렌젝션 + 트렌젝션-디 => merge

In [142]:
# 실습
# 공통 컬럼인 transaction_id 표기 안하고 진행
transaction_total = pd.merge(transaction, transaction_detail)
transaction_total.shape

# transaction_detail에 맞춰서 transaction 설정

(0, 10)

In [143]:
transaction_total
# detail_id, transaction_id 번호를 순번대로 배치되게 구성되었음

Empty DataFrame
Columns: [Unnamed: 0, detail_id, transaction_id, item_id, quantity, price, payment_date, customer_id, item_name, item_price]
Index: []

In [144]:
transaction_total.transaction_id.unique().shape
# 트렌젝션 1개당 트렌젝션 디테일이 1개 이상일수 있다!!

(0,)

In [145]:
# 앞쪽 컬럼 데이터는 중복성 x, 뒤쪽 데이터에서 중복성 등장
# 7144개의 데이터를 두고, 6786개 데이터를 중복적으로 교집합하여 붙이는 구조 정리
transaction_total = pd.merge(transaction_detail, transaction )
transaction_total.shape

(0, 10)

In [146]:
transaction_total
# detail_id, transaction_id 번호를 순번대로 배치되게 구성되었음

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id, detail_id, item_id, quantity, Unnamed: 0, item_name, item_price]
Index: []

#### 통합 트렌젝션 + 고객 마스터 => merge

- 트랜젝션 디테일 중심 데이터 배치 <- 고객 정보 세팅

In [147]:
# 실습 2 :  left join => 7144가 유지됨
df = pd.merge(transaction_total, dfs[0], on='customer_id', how='left')
df.shape

(0, 17)

In [148]:
df.head(1)

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id, detail_id, item_id, quantity, Unnamed: 0, item_name, item_price, customer_name, registration_date, email, gender, age, birth, pref]
Index: []

#### 통합 트렌젝션 + 아이템 마스터 => merge

In [149]:
# 실습 2
df = pd.merge(transaction_total, dfs[1], on='item_id', how='left')
df.shape

(0, 12)

In [150]:
df.head(1)

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id, detail_id, item_id, quantity, Unnamed: 0, item_name_x, item_price_x, item_name_y, item_price_y]
Index: []

#### 현재까지 작업 df 저장하기 (중간 결과물)

In [151]:
df.to_csv('/content/drive/MyDrive/2. 데이터분석/data/중간결과물.csv')

#### 검증

- 파생변수
  - cal_price = 아이템단가 * 주문수량
  - 주문서내에 총 결제액 == cal_price
  일치하는지 확인

In [153]:
# 실습
# 합치는 과정에서 주문 상세에 주문을 붙이는 과정 수행 -> 중복성이 나올 수도 있음
# 진행, 값이 상이하면 중복 여부 고민 필요
# 파생변수의 총합, 트랜젝션 원데이터의 매출 총액을 비교

# 거래 1건 => 디테일 => 제품별로 나눠져서 관리
# 파생변수 : cal_price = df.item_price * df.quantity
df['cal_price'] = df.item_price * df.quantity
df.head(1)

AttributeError: 'DataFrame' object has no attribute 'item_price'

In [154]:
# 합산
df.cal_price.sum() == transaction.price.sum()

AttributeError: 'DataFrame' object has no attribute 'cal_price'

In [155]:
# 고민 -> 왜 총합이 다른지 고민
# 트랜젝션(거래내역) 정보가 판매된 제품 기준으로 통합 -> 중복노출
# 금액이 더큼
df.price.sum()

np.float64(0.0)

### step3 데이터 클리닝

- 결측확인

In [156]:
df.info()
# RangeIndex: 7144 entries
# 개별 컬럼의 Non-Null Count => 7144 이면(동일하면) 결측 없음

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   transaction_id  0 non-null      object 
 1   price           0 non-null      float64
 2   payment_date    0 non-null      object 
 3   customer_id     0 non-null      object 
 4   detail_id       0 non-null      float64
 5   item_id         0 non-null      object 
 6   quantity        0 non-null      float64
 7   Unnamed: 0      0 non-null      float64
 8   item_name_x     0 non-null      object 
 9   item_price_x    0 non-null      float64
 10  item_name_y     0 non-null      object 
 11  item_price_y    0 non-null      int64  
dtypes: float64(5), int64(1), object(6)
memory usage: 132.0+ bytes


In [157]:
# 1개라 존재하면 0보다 큰값이 나옴, 1개면 1, ..
df.isna().sum(), df.isnull().sum()

(transaction_id    0
 price             0
 payment_date      0
 customer_id       0
 detail_id         0
 item_id           0
 quantity          0
 Unnamed: 0        0
 item_name_x       0
 item_price_x      0
 item_name_y       0
 item_price_y      0
 dtype: int64,
 transaction_id    0
 price             0
 payment_date      0
 customer_id       0
 detail_id         0
 item_id           0
 quantity          0
 Unnamed: 0        0
 item_name_x       0
 item_price_x      0
 item_name_y       0
 item_price_y      0
 dtype: int64)

### step4 월별 데이터로 집계 처리

- payment_date 항목 기준
  - 구매/판매일, 매출방생일
  - 날짜등 시간 정보 기준( 월단위, 일단위, 년간단위(데이터 부족), 요일단위, 시간단위)
    - 재고관리!! -> 수요예측 -> 생산, 주문,.. 수행

In [158]:
display(df.head(1))
display(df.tail(1))

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id, detail_id, item_id, quantity, Unnamed: 0, item_name_x, item_price_x, item_name_y, item_price_y]
Index: []

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id, detail_id, item_id, quantity, Unnamed: 0, item_name_x, item_price_x, item_name_y, item_price_y]
Index: []

In [159]:
df.dtypes # payment_date	object:문자열!!

transaction_id     object
price             float64
payment_date       object
customer_id        object
detail_id         float64
item_id            object
quantity          float64
Unnamed: 0        float64
item_name_x        object
item_price_x      float64
item_name_y        object
item_price_y        int64
dtype: object

In [160]:
# 데이터가 문자열인 경우 기준으로 해결

day = '2019-07-31 23:41:38'
# 실습1, day에서 월 정보 추출 -> 슬라이싱
#        날짜 포멧은 변경되지 않음으로 슬라이싱 사용

day[5:5+2]

'07'

In [161]:
day.split() # 공백 기준 분해 => 결과:년월일과 시분초가 분리

['2019-07-31', '23:41:38']

In [162]:
day.split()[0] # 년월일 선택(추출)

'2019-07-31'

In [163]:
day.split()[0].split('-') # 년, 월, 일 분해

['2019', '07', '31']

In [164]:
day.split()[0].split('-')[1] # 월 선택(추출)

'07'

In [165]:
# 실습2, df.payment_date에서 월정보 일괄 추출
# Series에서 데이터 하나씩 꺼내서 -> 전처리(월정보 추출) -> 다시 반환...
# apply()
# 데이터의 타입
# df.payment_date => 타입
# DF에서 컬럼 1개 선택(추출) -> 차원축소 -> Series
# type(df.payment_date)
df.payment_date.apply(lambda x: x[5:5+2]) # x = '2019-07-31 23:41:38'
# 02~07 : 7개월

Series([], Name: payment_date, dtype: object)

In [166]:
# 파생변수, df에 결제 정보중 월정보 별도 표현
df['pay_month'] = df.payment_date.c( lambda x: x[5:5+2] )

df.head(1)

AttributeError: 'Series' object has no attribute 'c'

In [167]:
# apply() -> str로 변경하여 동일한 결과가 나오게 처리
# 실습
# 데이터가 많아지고, 년정보에땋라 02가 여러번 등장한다면 구분값 필요
# 월 => 년월로 변경!!
df['pay_month'] = df.payment_date.str[:7].str.replace('-', '')


df.head(1)

Empty DataFrame
Columns: [transaction_id, price, payment_date, customer_id, detail_id, item_id, quantity, Unnamed: 0, item_name_x, item_price_x, item_name_y, item_price_y, pay_month]
Index: []

#### 원데이터가 문장열이나 그 형식이 시간 포멧인 경우, 년월 정보 추출

In [168]:
# 시간 포멧의 문자열 => datetime64 포멧변환
df['payment_date'] = pd.to_datetime( df.payment_date )

# 타입 확인
df.payment_date.dtype # 타입이 변경됨 => 시간함수로 추출가능함

dtype('<M8[ns]')

In [169]:
df.dtypes

transaction_id            object
price                    float64
payment_date      datetime64[ns]
customer_id               object
detail_id                float64
item_id                   object
quantity                 float64
Unnamed: 0               float64
item_name_x               object
item_price_x             float64
item_name_y               object
item_price_y               int64
pay_month                 object
dtype: object

In [170]:
# dt : datetime
df.payment_date.dt.month # 월정보 추출

Series([], Name: payment_date, dtype: int32)

In [171]:
# 장점: 원하는 시간정보를 포멧변경으로 추출가능함

# strftime(포멧):
# 시간(time)정보에서 특정 포멧(f)으로 추출 문자열(str)로 반환
# 시간 포멧: 'YYYY-mm-dd hh:mm:ss'
# '%Y%m : 년월 정보를 그대로 뽑아서, 이어붙여서 문자열로 반환, 포멧팅
df.payment_date.dt.strftime('%%m') # 년월정보 추출

Series([], Name: payment_date, dtype: object)

In [ ]:
df.head(1)

In [172]:
# pay_month는 오직 7개의 값만 가지고 있다
# 월'별' = => 집계(groupby) => '해당 월의 데이터 개수 확인'
# 실습
df.pay_month.value_counts()

# 월별 데이터의 개수가 대체적으로 비슷하다. 단, 최고(7), 최저(6) 사이에는
# 80여개 데이터 수 차이가 존재함 -> 보정> 층화(비율 유지하여 이후 처리)

Series([], Name: count, dtype: int64)

In [ ]:
# 다음 데이터프레임을 출력하시오
# 언제 피벗을 사용? -> 기준 컬럼이 범주형,
                    -> 인덱스, 컬럼, 데이터 구성에 대한 설명
                    -> XX'별' (피벗 or 집계)
# 월별, 아이템별 총 매출(월별 데이터는 합산) => 시각화 =? EDA 수행
# 인덱스 : pay_month
# 컬럼 : 아이템이름 : item_name
# 데이터(value) : cal_price -> sum 처리
# 실습
df.pivot_table( index   ='pay_month',
                columns ='item_name',
                values  ='cal_price',
                aggfunc ='sum' # 기본 평균
                )

In [ ]:
df.pivot_table( index   ='pay_month',
                columns ='item_name',
                values  ='cal_price',
                #aggfunc ='sum' # 기본 평균
                )

# 데이터 분석(생략)

# 모델 구축(생략)

# 시스템통합/산출물 작성(생략)